src/main/java/co/cask/http/AbstractHttpResponder.java

```diff
@@ -16,66 +16,84 @@
+    sendString(status, jsonString, new DefaultHttpHeaders().add(HttpHeaderNames.CONTENT_TYPE.toString(),
+                                                                "application/json"));
   }
 
   @Override
   public void sendString(HttpResponseStatus status, String data) {
-    sendString(status, data, null);
+    sendString(status, data, EmptyHttpHeaders.INSTANCE);
   }
 
   @Override
-  public void sendString(HttpResponseStatus status, String data, @Nullable Multimap<String, String> headers) {
+  public void sendString(HttpResponseStatus status, String data, HttpHeaders headers) {
     if (data == null) {
       sendStatus(status, headers);
       return;
     }
-    try {
-      ByteBuf buffer = Unpooled.wrappedBuffer(Charsets.UTF_8.encode(data));
-      sendContent(status, buffer, "text/plain; charset=utf-8", headers);
-    } catch (Exception e) {
-      throw Throwables.propagate(e);
-    }
+    ByteBuf buffer = Unpooled.wrappedBuffer(StandardCharsets.UTF_8.encode(data));
+    sendContent(status, buffer, addContentTypeIfMissing(new DefaultHttpHeaders().setAll(headers),
+                                                        "text/plain; charset=utf-8"));
   }
 
   @Override
   public void sendStatus(HttpResponseStatus status) {
-    sendContent(status, null, null, ImmutableMultimap.<String, String>of());
+    sendContent(status, Unpooled.EMPTY_BUFFER, EmptyHttpHeaders.INSTANCE);
   }
 
   @Override
-  public void sendStatus(HttpResponseStatus status, @Nullable Multimap<String, String> headers) {
-    sendContent(status, null, null, headers);
+  public void sendStatus(HttpResponseStatus status, HttpHeaders headers) {
+    sendContent(status, Unpooled.EMPTY_BUFFER, headers);
   }
 
   @Override
-  public void sendByteArray(HttpResponseStatus status, byte[] bytes, @Nullable Multimap<String, String> headers) {
+  public void sendByteArray(HttpResponseStatus status, byte[] bytes, HttpHeaders headers) {
     ByteBuf buffer = Unpooled.wrappedBuffer(bytes);
-    sendContent(status, buffer, "application/octet-stream", headers);
+    sendContent(status, buffer, headers);
+  }
+
+  @Override
+  public void sendBytes(HttpResponseStatus status, ByteBuffer buffer, HttpHeaders headers) {
+    sendContent(status, Unpooled.wrappedBuffer(buffer), headers);
   }
 
   @Override
-  public void sendBytes(HttpResponseStatus status, ByteBuffer buffer, @Nullable Multimap<String, String> headers) {
-    sendContent(status, Unpooled.wrappedBuffer(buffer), "application/octet-stream", headers);
+  public void sendFile(File file) throws IOException {
+    sendFile(file, EmptyHttpHeaders.INSTANCE);
   }
 
+  @Override
+  public ChunkResponder sendChunkStart(HttpResponseStatus status) {
+    return sendChunkStart(status, EmptyHttpHeaders.INSTANCE);
+  }
+
+  protected final HttpHeaders addContentTypeIfMissing(HttpHeaders headers, String contentType) {
```

Comment:
if `headers` already contains the key, should we check equality with the existing value?

Reply:
hum... not for this use case. This method is to add the content-type header if it is missing. If it is already set by the caller, then no change will be made